compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmDesugarPhase.java

```diff
@@ -249,208 +249,218 @@ private JvmDesugarPhase() {
+            function.workerName = names.fromString(getInitialIdString(function.workerName.value,
+                                                                      encodedVsInitialIds));
         }
     }
 
     private static void replaceEncodedAttachedFunctionIdentifiers(List<BAttachedFunction> functions, Names names,
-                                                                  HashMap<String, String> originalIdentifierMap) {
+                                                                  HashMap<String, String> encodedVsInitialIds) {
         for (BAttachedFunction function : functions) {
-            function.funcName = names.fromString(originalIdentifierMap.get(function.funcName.value));
+            function.funcName = names.fromString(getInitialIdString(function.funcName.value, encodedVsInitialIds));
         }
     }
 
     private static void replaceEncodedGlobalVariableIdentifiers(List<BIRNode.BIRGlobalVariableDcl> globalVars,
                                                                 Names names,
-                                                                HashMap<String, String> originalIdentifierMap) {
+                                                                HashMap<String, String> encodedVsInitialIds) {
         for (BIRNode.BIRGlobalVariableDcl globalVar : globalVars) {
             if (globalVar == null) {
                 continue;
             }
-            String originalGlobalVarName = originalIdentifierMap.get(globalVar.name.value);
-            if (originalGlobalVarName != null) {
-                globalVar.name = names.fromString(originalGlobalVarName);
-            }
+            globalVar.name = names.fromString(getInitialIdString(globalVar.name.value, encodedVsInitialIds));
         }
     }
 
-    private static String encodeFunctionIdentifier(String functionName, HashMap<String, String> originalIdentifierMap) {
-        String encodedString = IdentifierUtils.encodeFunctionIdentifier(functionName);
-        originalIdentifierMap.putIfAbsent(encodedString, functionName);
+    private static String encodeFunctionIdentifier(String initialIdString,
+                                                   HashMap<String, String> encodedVsInitialIds) {
+        if (encodedVsInitialIds.containsKey(initialIdString)) {
+            return initialIdString;
+        }
+        String encodedString = IdentifierUtils.encodeFunctionIdentifier(initialIdString);
+        encodedVsInitialIds.putIfAbsent(encodedString, initialIdString);
         return encodedString;
     }
 
-    private static String encodeNonFunctionIdentifier(String pkgName, HashMap<String, String> originalIdentifierMap) {
-        String encodedString = IdentifierUtils.encodeNonFunctionIdentifier(pkgName);
-        originalIdentifierMap.putIfAbsent(encodedString, pkgName);
+    private static String encodeNonFunctionIdentifier(String encodedIdString,
```

Comment:
The variable names seems a bit unclear as both `encodedIdString` , `encodedString` contains `encoded` part. Shall we rename the parameter?